### DISCUSSION TITLE
Cannot convert GIF URL to Bitmap

### DISCUSSION THREAD

**Discussion by briancatraguna** [coil-kt/coil#976] [Q&A]
**Describe the bug**
I'm trying to convert a GIF url to bitmap using coil. However the result that I am getting is a `MovieDrawable` instead of a `BitmapDrawable`. This is how we're trying to do it. The image url data that I'm working with is in the form of a local storage address that's why we're putting it like that.

```KT
  private fun getBitmapFromUrl(
        context: Context,
        imageUrl: String
    ): Bitmap =  runBlocking {
        val request = ImageRequest.Builder(context)
            .data("file://$imageUrl")
            .build()
        val result = (imageLoader.execute(request))
        val drawable = (result as SuccessResult).drawable
        drawable.toBitmap()
    }
```

The reason I'm trying to do it is because we needed to integrate coil with an in app messaging service from Braze. To integrate it, I needed to implement an external image library to convert an image url to a bitmap as outlined in this document: https://www.braze.com/docs/developer_guide/platform_integration_guides/android/in-app_messaging/customization/#dismiss-modal-on-outside-tap

In the documentation, it says that I can use any image library, and the example uses glide. However since our team uses coil as our image library, we wanted to use that as our external image library to integrate it. This is the example when we use Glide:
```KT
private fun getBitmapFromUrl(
        context: Context,
        imageUrl: String,
        viewBounds: BrazeViewBounds?
    ): Bitmap? {
        try {
            return Glide.with(context)
                .asBitmap()
                .apply(mRequestOptions)
                .load(imageUrl).submit().get()
        } catch (e: Exception) {
            Timber.e(e, TAG, "Failed to retrieve bitmap at url: %s", imageUrl)
        }
        return null
    }
```
The idea is to convert a gif url to a bitmap. With Glide, it works perfectly. The gif shows up when using Glide:

 <img src="https://user-images.githubusercontent.com/32363208/140689532-1ef961db-8c81-4eec-bd3c-52a8ff11543f.jpg" width=500/>

But this is what shows up when using coil:

<img src="https://user-images.githubusercontent.com/32363208/140516432-3fec2009-5ed1-4d9c-bcf1-a6a19aacaa10.png" width=400/>

So I'm not sure how to convert an image url to bitmap properly using coil. Is there anyone who understands how to solve this issue?
**Version**
Library version:
```
implementation "io.coil-kt:coil:1.0.0-rc2"
```

**Comment by colinrtwhite** [comment]
Hmm `drawable.toBitmap()` should convert the `MovieDrawable` to bitmap that displays the first frame. I'd try updating Coil to the latest since you're using an old library version. I'd also try setting `GifDecoder(enforceMinimumFrameDelay = true)` when adding the decoder to your component registry. I'd also try using `ImageDecoderDecoder` as well.

**Comment by laithnurie** [comment]
Hi there,

I am interested in this issue, I tried doing the below but I was only able to get the 3 frames out of 8 in the `invalidateDrawable` callback and that's it, can someone advice on what to do ?
```
movieDrawable.registerAnimationCallback(object : Animatable2Compat.AnimationCallback() {
        override fun onAnimationStart(drawable: Drawable) {
          super.onAnimationStart(drawable)
          Log.v("lnln", "onAnimationStart")
          bitmapList.add(drawable.toBitmap())
        }

        override fun onAnimationEnd(drawable: Drawable) {
          super.onAnimationEnd(drawable)
          Log.v("lnln", "onAnimationEnd")
          onImageLoaded(bitmapList)
        }
      })
      movieDrawable.callback = object : Drawable.Callback {
        override fun invalidateDrawable(who: Drawable) {
          Log.v("lnln", "invalidateDrawable")
        }

        override fun scheduleDrawable(who: Drawable, what: Runnable, `when`: Long) {
          Log.v("lnln", "scheduleDrawable")
        }

        override fun unscheduleDrawable(who: Drawable, what: Runnable) {
          Log.v("lnln", "unscheduleDrawable")
        }
      }
      movieDrawable.start()
    }
    ```
